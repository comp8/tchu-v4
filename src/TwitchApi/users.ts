import Config from "../config";

interface ITwitchUserResponse {
  data: ITwitchUser[];
}
interface ITwitchUser {
  id: string;
  login: string;
  display_name: string;
  description: string;
  broadcaster_type: string;
  offline_image_url: string;
  profile_image_url: string;
  type: 'staff' | 'admin' | 'global_mod' | '';
  view_count: number;
  email?: string;
  created_at: string;
}

interface IFetchUsersOptions {
  ids?: string[];
  logins?: string[];
  access_token: string;
}

export async function fetchUsers({ ids, logins, access_token }: IFetchUsersOptions): Promise<ITwitchUserResponse> {
  const url = 'https://api.twitch.tv/helix/users';

  ids = ids || [];
  logins = logins || [];

  const params = [...ids.map(e => 'id=' + e), ...logins.map(e => 'login=' + e)];
  if (params.length <= 0) {
    return null;
  }

  try {
    const response = await fetch(url + '?' + params.join('&'), {
      headers: {
        'Authorization': 'Bearer ' + access_token,
        'Client-Id': Config.Twitch.clientId
      },
      method: 'GET',
    });

    if (response.status === 200) {
      return await response.json();
    }
    console.error(response);
  }
  catch (ex) {
    console.error(ex);
  }
  return null;
}