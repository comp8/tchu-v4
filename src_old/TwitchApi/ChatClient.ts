import React, { useEffect, useState } from 'react';
import { useDispatch } from 'react-redux';
import tmi from 'tmi.js';
import Config from '../config';
import Actions from '../store/actions';

const connection_delay = 600;

interface ChatClientProps {
  channels: string[];
  username?: string;
  password?: string;
}

export class ChatClient extends tmi.Client {

  constructor(props: ChatClientProps) {
    super({
      options: {
        // debug: process.env.NODE_ENV === 'development',
        clientId: Config.Twitch.clientId
      },
      identity: props.username && props.password ? {
        username: props.username,
        password: 'oauth:' + props.password,
      } : undefined,
      channels: [
        ...props.channels
      ],
    });
  }

  destroy() {
    this.removeAllListeners();
    const readyState = this.readyState();
    if (readyState !== 'CLOSED') {
      try {
        this.disconnect();
        console.log(readyState, this.readyState());
      } catch (ex) { }
    }
  }
}

export type IChatClientListener<TEventName extends keyof tmi.Events> = tmi.Events[TEventName];

export function useChatClient(opts: { channel: string, name?: string, pass?: string }, deps: React.DependencyList = []) {

  const { channel, name: username, pass: password } = opts;

  const dispatch = useDispatch();
  const [client, setClient] = useState<ChatClient>(null);

  useEffect(() => {
    const cc = new ChatClient(Object.assign({
      channels: [channel]
    },
      username && password ? { username, password } : null
    ));

    cc.on('chat', (channel, userstate, message, self) => {
      dispatch(Actions.Chats.Append({ chat: { channel, userstate, message } }));
      dispatch(Actions.Users.Update({ chat: { channel, userstate, message } }));
    });
    cc.on('clearchat', () => {
      dispatch(Actions.Chats.Clear());
    });

    const delayedConnect = (delay: number) => setTimeout(() => cc.connect(), delay);
    const timerId = delayedConnect(connection_delay);

    setClient(cc);

    return () => {
      clearTimeout(timerId);
      cc.destroy();
      setClient(null);
    };
  }, [channel, username, password, ...deps]);

  return client;
}
