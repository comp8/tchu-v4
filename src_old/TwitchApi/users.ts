import Config from "../config";

interface ITwitchUserResponse {
  data: ITwitchUser[];
}
interface ITwitchUser {
  id: string;
  login: string;
  display_name: string;
  description: string;
  broadcaster_type: string;
  offline_image_url: string;
  profile_image_url: string;
  type: 'staff' | 'admin' | 'global_mod' | '';
  view_count: number;
  email?: string;
  created_at: string;
}

interface IFetchUsersOptions {
  ids?: string[];
  logins?: string[];
  access_token: string;
}

export async function fetchUsers({ ids, logins, access_token }: IFetchUsersOptions): Promise<ITwitchUserResponse> {
  const url = 'https://api.twitch.tv/helix/users';

  ids = ids || [];
  logins = logins || [];

  const params = [...ids.map(e => 'id=' + e), ...logins.map(e => 'login=' + e)];
  if (params.length <= 0) {
    return null;
  }

  try {
    const response = await fetch(url + '?' + params.join('&'), {
      headers: {
        'Authorization': 'Bearer ' + access_token,
        'Client-Id': Config.Twitch.clientId
      },
      method: 'GET',
    });

    if (response.status === 200) {
      return await response.json();
    }
    console.error(response);
  }
  catch (ex) {
    console.error(ex);
  }
  return null;
}

interface IFetchUsersFollowsOptions {
  from_id?: string;
  to_id?: string;
  first?: number;
  after?: string;
  access_token: string;
}
interface ITwitchUserFollowsResponse {
  total: number;
  pagination?: {
    cursor?: string;
    [key: string]: string;
  };
  data: ITwitchUserFollows[];
}
interface ITwitchUserFollows {
  followed_at: string;
  from_id: string;
  from_login: string;
  from_name: string;
  to_id: string;
  to_login: string;
  to_name: string;
}

export async function fetchUsersFollows({ from_id, to_id, after, first, access_token }: IFetchUsersFollowsOptions): Promise<ITwitchUserFollowsResponse> {
  const url = 'https://api.twitch.tv/helix/users/follows';

  if (!from_id && !to_id) return null;

  const params = new URLSearchParams();

  from_id && params.set('from_id', from_id);
  to_id && params.set('to_id', to_id);
  after && params.set('after', after);
  first && params.set('first', first.toString());

  try {
    const response = await fetch(url + '?' + params.toString(), {
      headers: {
        'Authorization': 'Bearer ' + access_token,
        'Client-Id': Config.Twitch.clientId
      },
      method: 'GET',
    });

    if (response.status === 200) {
      return await response.json();
    }
    console.error(response);
  }
  catch (ex) {
    console.error(ex);
  }
  return null;
}