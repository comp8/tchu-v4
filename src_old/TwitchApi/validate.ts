interface ITwitchValidateSuccess {
  client_id: string;
  login: string;
  scopes: string[];
  user_id: string;
  expires_in: number;
}

interface ITwitchValidateFailed {
  status: number;
  message: string;
}

type ITwitchValidateResponse = ITwitchValidateSuccess & ITwitchValidateFailed;

export async function validateTwitchAccessToken(token: string): Promise<ITwitchValidateResponse> {
  const url = 'https://id.twitch.tv/oauth2/validate';
  const r = await fetch(url, {
    headers: {
      'Authorization': 'OAuth ' + token
    }
  });
  return await r.json() as ITwitchValidateResponse;
}
