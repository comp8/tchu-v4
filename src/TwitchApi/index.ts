import React, { useEffect, useState } from 'react';
import { useDispatch } from 'react-redux';
import Actions from '../Store/actions';
import { BadgeSets, fetchBadgeSets } from './badges';
import { fetchUsers } from './users';
import { validateTwitchAccessToken } from './validate';

export * from './users';
export * from './badges';
export * from './validate';
export * from './ChatClient';


export function useTwitchValidation(deps: React.DependencyList = []) {
  const [loginInfo, setLoginInfo] = useState<{ login?: string, user_id?: string }>({});
  const { login, user_id } = loginInfo;

  const token = localStorage.getItem('token');

  useEffect(() => {
    let timerId: number | NodeJS.Timeout = null;

    async function validate() {
      const r = await validateTwitchAccessToken(token);
      if (r?.user_id && r.login) {
        if (r.expires_in) {
          clearTimeout(timerId);
          const interval = (r.expires_in - Date.now()) / 2;
          timerId = setTimeout(() => validate(), interval);
        }
        setLoginInfo({ login: r.login, user_id: r.user_id });
      }
      if (!r || r.status > 400) {
        console.error('token validation error', r);
      }
    }
    validate();

    return () => {
      clearInterval(timerId);
    };
  }, [token, ...deps]);

  return { login, user_id, access_token: token };
}

interface IBroadcasterInfo {
  uid: string;
  uname: string;
  dname: string;
  profile: string;
};
const initialBroadcasterInfo: IBroadcasterInfo = {
  uid: null,
  uname: null,
  dname: null,
  profile: null,
};

export function useBroadcasterInfo({ channel, access_token }: { channel: string, access_token: string }, deps: React.DependencyList = []) {
  const [info, setInfo] = useState<IBroadcasterInfo>(initialBroadcasterInfo);
  const dispatch = useDispatch();

  useEffect(() => {
    if (channel) {
      fetchUsers({ logins: [channel], access_token }).then(({ data }) => {
        if (data?.[0]?.id) {
          const { display_name: dname, login: uname, id: uid, profile_image_url: thumb } = data[0];
          const obj = { dname, uid, uname, token: access_token };
          setInfo({ uid, dname, uname, profile: thumb });
          dispatch(Actions.Channel.Update({ data: obj }));
        }
      });
    }
    else {
      setInfo(initialBroadcasterInfo);
      dispatch(Actions.Channel.Update({ data: { dname: null, uid: null, uname: null } }));
    }
  }, [channel, ...deps]);

  return info;
}

export function useBadgeSets(user_id: string, deps: React.DependencyList = []) {

  const [badgeSets, setBadgeSets] = useState<BadgeSets>({});

  useEffect(() => {
    user_id && fetchBadgeSets(user_id).then(r => setBadgeSets(r?.badge_sets || {}));
  }, [user_id, ...deps]);

  return badgeSets;
}
