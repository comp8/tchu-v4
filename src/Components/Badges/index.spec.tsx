import React from 'react';
import renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';
import { Badges } from "."
import { BadgeSets } from '../../TwitchApi';


describe('Badges', () => {

  let mockUseContext: ReturnType<typeof jest.fn>;
  let realUseContext: typeof React.useContext;

  beforeEach(() => {
    realUseContext = React.useContext;
    mockUseContext = React.useContext = jest.fn();
  });

  afterEach(() => {
    React.useContext = realUseContext;
  })

  test('', () => {
    const comp = renderer.create(<Badges badges={{}} />);
    expect(comp.toJSON()).toMatchSnapshot();
  });

  test('', () => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render(<Badges badges={{ subscriber: '6', testBadge: '1', otherBadge: '2' }} />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
    expect(result.props.children.length).toBe(3);
  });
})