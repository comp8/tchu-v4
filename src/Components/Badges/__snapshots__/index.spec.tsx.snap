// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badges  1`] = `
<span
  className="badges-wrapper"
/>
`;

exports[`Badges  2`] = `
<span
  className="badges-wrapper"
>
  <Badge
    badgeId="subscriber"
    version="6"
  />
  <Badge
    badgeId="testBadge"
    version="1"
  />
  <Badge
    badgeId="otherBadge"
    version="2"
  />
</span>
`;
